File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
          unordered_map<char,int>lastseen;
        int left = 0 ;
        int maxlength = 0;

        for ( int right = 0 ; right < s.size() ; right ++ ){
            char c = s[right];

            if ( lastseen.count(c) && lastseen[c] >= left ){
                left = lastseen[c] + 1;
            }

            lastseen[c] = right;
            maxlength = max ( maxlength , right - left + 1);
        }

        return maxlength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastSeen = new HashMap<>();
        int maxLength = 0;
        int start = 0;

        for (int end = 0; end < s.length(); ++end) {
            char ch = s.charAt(end);

            if (lastSeen.containsKey(ch) && lastSeen.get(ch) >= start) {
                start = lastSeen.get(ch) + 1;
            }

            lastSeen.put(ch, end);
            maxLength = Math.max(maxLength, end - start + 1);
        }

        return maxLength;
    }
}